General update

SolidifiedRay opened the pull request:
Make some changes due to the latest frontend and Python 3.

See commit messages for the details.

SolidifiedRay committed:
Replace request.is_xhr

Request.is_xhr has been deprecated as it is not reliable, according to issue #1077 in Werkzeug(WSGI) repo

Signed-off-by: SolidifiedRay <yc3346@nyu.edu>

SolidifiedRay committed:
Fix docstring

Signed-off-by: SolidifiedRay <yc3346@nyu.edu>

SolidifiedRay committed:
Replace StringIO with ByteIO

Flask on python3 won't work with StringIO as send_file only expects to be given binary file-like objects and it wouldn't know how to encode the string to bytes, according to issue #3358 in Flask repo

Signed-off-by: SolidifiedRay <yc3346@nyu.edu>

lukpueh commented:
Will this still work on Python 2? If not, we should mention it maybe in the commit message and also change the requirements section in the README.md.

lukpueh commented:
@SolidifiedRay, would you mind taking a look at #39, #40, #41, #42 and see if updating those breaks anything?

SolidifiedRay commented:
> Will this still work on Python 2? If not, we should mention it maybe in the commit message and also change the requirements section in the README.md.

This still work on Python 2. I will take a look at other updates soon.

SolidifiedRay commented:
> @SolidifiedRay, would you mind taking a look at #39, #40, #41, #42 and see if updating those breaks anything?

They look good to me. I ran them locally and did some quick tests. No errors show up. @lukpueh

lukpueh commented:
> They look good to me. I ran them locally and did some quick tests. No errors show up. @lukpueh

Thanks for checking. Let's merge them in then.

adityasaky commented on wizard.py:
```diff
@@ -1044,8 +1044,8 @@ def download_layout():
-  layout_fp = StringIO.StringIO()
-  layout_fp.write("{}".format(layout_metadata))
+  layout_fp = io.BytesIO()
+  layout_fp.write(("{}".format(layout_metadata)).encode())
```
Could you explain this bit here? Is that a py2 -> py3 change, to encode?

SolidifiedRay replied:
Flask.send_file will send blank files when using io.StringIO in python 3. 
Someone also opens an issue about this in Flask repository, and one of the members of Flask reply:

> The fact that we support BytesIO is not an indication that we should support StringIO. send_file only expects to be given binary file-like objects. The same issue happens when passing an open file in text mode. The function doesn't deal with encoding, so it wouldn't know how to encode the string to bytes, and I think that should remain up to the user code calling it.

> Determining if a given file-like object reads bytes appears to be non-trivial, especially with compatibility with Python 2's file. It's probably not worth the overhead in general for send_file. As a compromise, I can add a mention of the mode to the docs, and raise an error if it's an instance of io.TextIOBase, which should cover most cases including the one given here.

Here is the [link ](https://github.com/pallets/flask/issues/3358) for this issue.

lukpueh replied:
Thanks for the background information. For future reference, I find it very useful to include information like that (to a reasonable extent) in the commit message. :)

SolidifiedRay replied:
I see. I have updated the commit messages to include more information now.

SolidifiedRay commented:
I go back to have another look at `Request.is_xhr` and find Werkzeug(WSGI) repo also has an issue for this.
According to [issue #1077](https://github.com/pallets/werkzeug/issues/1077) in werkzeug repo, werkzeug removed Request.is_xhr as it is not reliable. Since Flask depends on the Werkzeug toolkit, `Request.is_xhr` is also no longer supported in Flask.
Here is a quote from issue # 1077:

> The Request.is_xhr was introduced 9 years ago. It detects from a header called X-Requested-With which is added by JavaScript libraries.
> 
> The world is different now. jQuery is losing its popularity; and fetch is shining. Many would choose to use fetch instead of XMLHttpRequest nowadays. Popular ajax libraries such as axios has no built-in X-Requested-With header.
> 
> X-Requested-With is not a part of any spec, it is not reliable. It is the time to deprecate it.

Maybe we should also remove the check for `XMLHttpRequest`? What do you think? @lukpueh

SolidifiedRay committed:
Edit in-toto version

install in-toto directly instead of installing in an editable way and a dev version

Signed-off-by: SolidifiedRay <yc3346@nyu.edu>

adityasaky reviewed (approved):
LGTM, thanks @SolidifiedRay!

lukpueh commented on requirements.txt:
```diff
@@ -1,4 +1,4 @@
 Flask-PyMongo
 Flask-WTF
--e git://github.com/in-toto/in-toto.git@v0.2.3.dev5#egg=in-totov0.2.3.dev5
+in-toto
```
I'm against un-pinning in-toto:
See https://github.com/in-toto/layout-web-tool/pull/37#discussion_r450067804 and https://www.promptworks.com/blog/pin-all-dependencies.

adityasaky replied:
Very interesting, and I'd missed the previous comment, my bad.

Should we pin all of them to their latest versions in this PR?

lukpueh replied:
Yes, it's probably a good idea to pin at least major version of the others too.

SolidifiedRay committed:
Edit installation requirements

Signed-off-by: SolidifiedRay <yc3346@nyu.edu>

SolidifiedRay committed:
pin version of other requirements

Signed-off-by: SolidifiedRay <yc3346@nyu.edu>

adityasaky reviewed (approved):
LGTM, thanks @SolidifiedRay!

lukpueh commented on wizard.py:
```diff
@@ -362,7 +362,7 @@ def ajax_flash_messages(response):
   """
 
-  if (request.is_xhr and
+  if (request.headers.get("X-Requested-With") == "XMLHttpRequest"  and
```
IIUC you are re-implementing the check that the deprecated `request.is_xhr` used to perform, right?

Did you have a chance to check if the jquery ajax call we use still sets that header, so that our `ajax_flash_messages` hook still works?

SolidifiedRay replied:
Thanks again for the review @lukpueh ! I am a little bit confused about the `ajax_flash_messages hook` and I am not sure what a hook does. Could you please explain it a bit more?

SolidifiedRay replied:
Thanks a ton for the review @lukpueh !!

I checked the `ajax_flash_messages` hook and unfortunately something is not right.
For example, when I upload the functionary's public key in the functionaries page, the flash messages show up only after I manually refresh the page. When I try to remove the functionary that already has a pubkey, I also have to refresh the page to see the change and the flash message. I am not sure if there is something wrong with the ajax call or the `ajax_flash_messages` hook, but I will try to find up why this happened.

SolidifiedRay replied:
As a follow-up, I will work on [rule generation](https://github.com/in-toto/layout-web-tool/issues/48#issuecomment-671955671) first, which seems to have a higher priority than this problem.

SolidifiedRay replied:
I found `ajax_flash_messages` working again.

I check the jquery ajax call with the browser's network console, and it does include the `X-Requested-With` header with the value `XMLHttpRequest`. So I believe there is no problem with the code here. I doubt that it failed previously is because my Virtual Machine didn't run the WSGI server correctly.

lukpueh replied:
I'll merge then. Thanks for the updates!
